wandb/sdk/launch/runner/kubernetes.py

```diff
@@ -0,0 +1,222 @@
+            raise LaunchError("Resource overrides not supported for multiple containers. Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+
+        containers[0]['name'] = resource_args.get("container_name", containers[0].get('name', 'launch'))
+        container_resources = containers[0].get('resources', {})
+        if resource_args.get('resource_requests'):
+            container_resources['requests'] = resource_args.get("resource_requests")
+        if resource_args.get('resource_limits'):
+            container_resources['limits'] = resource_args.get('resource_limits')
+        if container_resources:
+            containers[0]['resources'] = container_resources
+        # todo: args and env vars would be added here, need to figure out what kind of overrides we want
+
+        pod_spec['restartPolicy'] = resource_args.get('pod_restart_policy', 'Never')
+        if resource_args.get('pod_preemption_policy'):
+            pod_spec['preemptionPolicy'] = resource_args.get("pod_preemption_policy")
+        if resource_args.get('node_name'):
+            pod_spec['nodeName'] = resource_args.get('node_name')
+        if resource_args.get('node_selectors'):
+            pod_spec['nodeSelectors'] = resource_args.get('node_selectors')
+
+        validate_docker_installation()
+        entry_point = launch_project.get_single_entry_point()
+        docker_args: Dict[str, Any] = self.backend_config[PROJECT_DOCKER_ARGS]
+        if docker_args and list(docker_args) != ['docker_image']:
+            wandb.termwarn(
+                "Docker args are not supported for Kubernetes. Not using docker args"
+            )
+
+        image = resource_args.get("image")
+        if image:
+            if launch_project.docker_image and image != launch_project.docker_image:
+                raise LaunchError("Conflicting Docker images specified in launch and resource args.")
+            if len(containers) > 1:
+                raise LaunchError("Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+        user_provided_image = image or launch_project.docker_image
+        if user_provided_image:
+            pull_docker_image(user_provided_image)
+            containers[0]['image'] = user_provided_image
+        else:
+            registry = resource_args.get("registry")
+            if registry is None:
+                # allow local registry usage for eg local clusters but throw a warning
+                wandb.termlog("Warning: No Docker registry specified. Image will be hosted on local registry, which may not be accessible to your training cluster.")
+
+            image_uri = construct_local_image_uri(launch_project)
+            if registry:
+                image_uri = os.path.join(registry, image_uri)
+            generate_docker_image(self._api, launch_project, image_uri, entry_point, docker_args, runner_type="kubernetes")
```

Comment:
I don't see any changes to generate_docker_image in this PR, does it do something different for kubernetes?

Reply:
nope although after this PR i need to add generalized command/entrypoint functionality